sdk/keyvault/azure-keyvault-keys/azure/keyvault/keys/crypto/_models.py

```diff
@@ -2,11 +2,176 @@
+}
+OAEP_MAP = {
+    SHA1: EncryptionAlgorithm.rsa_oaep,
+    SHA256: EncryptionAlgorithm.rsa_oaep_256
+}
+PSS_MAP = {
+    SignatureAlgorithm.rs256: SignatureAlgorithm.ps256,
+    SignatureAlgorithm.rs384: SignatureAlgorithm.ps384,
+    SignatureAlgorithm.rs512: SignatureAlgorithm.ps512,
+}
+
+
+class ManagedRsaKey(RSAPrivateKey):
+    """An `RSAPrivateKey` implementation based on a key managed by Key Vault.
+
+    :param str key_name: The name of the key vault key to use.
+    :param key_client: The client that will be used to communicate with Key Vault.
+    :type key_client: KeyClient
+    """
+
+    def __init__(self, key_name: str, key_client: "KeyClient") -> None:  # pylint:disable=unused-argument
+        self._key_name: str = key_name
+        self._key_client: "KeyClient" = key_client
+        self._crypto_client: "CryptographyClient" = key_client.get_cryptography_client(key_name)
+
+    def decrypt(self, ciphertext: bytes, padding: AsymmetricPadding) -> bytes:
+        """Decrypts the provided ciphertext.
+
+        :param bytes ciphertext: Encrypted bytes to decrypt.
+        :param padding: The padding to use. Supported paddings are `OAEP` and `PKCS1v15`. For `OAEP` padding, supported
+            hash algorithms are `SHA1` and `SHA256`. The only supported mask generation function is `MGF1`. See
+            https://learn.microsoft.com/azure/key-vault/keys/about-keys-details for details.
+        :type padding: AsymmetricPadding
+
+        :returns: The decrypted plaintext, as bytes.
+        :rtype: bytes
+        """
+        if isinstance(padding, OAEP):
+            # Public algorithm property was only added in https://github.com/pyca/cryptography/pull/9582
+            # _algorithm property has been available in every version of the OAEP class, so we use it as a backup
+            try:
+                algorithm = padding.algorithm
+            except AttributeError:
+                algorithm = padding._algorithm
+            mapped_algorithm = OAEP_MAP.get(type(algorithm))
+            if mapped_algorithm is None:
+                raise ValueError(f"Unsupported algorithm: {algorithm.name}")
+    
+            # Public mgf property was added at the same time as algorithm
+            try:
+                mgf = padding.mgf
+            except AttributeError:
+                mgf = padding._mgf
+            if not isinstance(mgf, MGF1):
+                raise ValueError(f"Unsupported MGF: {mgf}")
+        elif isinstance(padding, PKCS1v15):
+            mapped_algorithm = EncryptionAlgorithm.rsa1_5
+        else:
+            raise ValueError(f"Unsupported padding: {padding.name}")
+        result = self._crypto_client.decrypt(mapped_algorithm, ciphertext)
+        return result.plaintext
+
+    @property
+    def key_size(self) -> int:
+        """The bit length of the public modulus.
+
+        :returns: The key's size.
+        :rtype: int
+        """
+        key = self._key_client.get_key(self._key_name)
```

Comment:
Do you attempt to cache this in memory already? You should. It's a waste of bandwidth otherwise and counts against the low rate limits. More than anything, the `CryptographyClient` was originally designed as a mitigation for low rate limits, at least for public key operations.

Reply:
Yep, we can definitely borrow the client's caching to save on fetches. As @arsatiki pointed out in the linked issue, we may also want to link instance creation to the `CryptographyClient` with a method on that client anyway, which could make things easier

Reply:
That's what I do in .NET: https://github.com/Azure/azure-sdk-for-net/blob/8a110f564fb2a8982d4d7b7128978c80a22680bf/sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs#L1642